### PR TITLE
add inventory plugin for Proxmox VE

### PR DESCRIPTION
**Pull request by xoxys**
##### SUMMARY
Add a new inventory plugin for Proxmox VE. The source code is partially based on https://github.com/xezpeleta/Ansible-Proxmox-inventory and was rewritten to use `proxmoxer` library instead of native API calls.

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
inventory/proxmox.py

### CODE REVIEW

**Comment by xoxys** [comment]
Damn... I realized it too late, there is already an open PR to add a Proxmox inventory. Maybe this could be closed, sorry for the duplicate

**Comment by xoxys** [comment]
I know a lot of work has already been put into the other PR (https://github.com/ansible-collections/community.general/pull/545) but IMO using the same API library already used in the `proxmox_` modules is the better approach in general. I've added some basic tests for now, let me know what do you think and ho to proceed. Feel free to close this PR and work with the existing one further, fine for me as well.

**Comment by Thulium-Drake** [comment]
Not sure if I agree, or disagree:

1. This module is consistent with the other proxmox modules for Ansible as it is using the same PVE API library, however, I feel that library's added value is really low (at the cost of another dependency).
2. I don't really see pip as a reliable source of packages, installing them on RHEL based systems can be an issue (Debian too for that matter, but they package a lot of stuff that's in pip)

So, given the low added value of proxmoxer (which itself in turn uses requests) I feel a direct API approach is better (I only wrote my other plugin, proxmox_snap, based on the other modules, to get it up and running quicker, but given enough interest I'd like to look into rewriting them without Proxmoxer)

Anyone want to chime in on this? :-)

P.S. I might look at your tests for inspiration for my modules, making unit tests is not a skill I master as of now..

**Comment by xoxys** [comment]
Thats fine dont want to start a "better solution" war :) 

There are some features I would like to see:

- exclude options e.g. for vmid or status
- IP address gathering for qemu agents and containers
- configurable default group all hosts applied to

**Comment by xoxys** [comment]
The main advantage of using libraries is the abstraction and potential duplicate code. Lets say there are 5 proxmox modules and 2 other Plugins and everyone is writing own API requests, every of them is duplicating the same code + everyone need to care of API changes

**Comment by Thulium-Drake** [comment]
> Thats fine dont want to start a "better solution" war :)
> 
Please, no, there's already enough flamewars on the internet ;-)
> There are some features I would like to see:
> 
>     * exclude options e.g. for vmid or status
Do you mean you wish to exclude certain vm states from the inventory? I'd say it's better to assign them groups per state, so you can still see them in the inventory (I see that you also assign running/stopped state groups)
> 
>     * IP address gathering for qemu agents and containers
That should be doable, but it might be tricky, which IP would you connect to, if the system has multiple addresses, but not all of them are management? You can't really see that on the outside. I feel that changing the VM name to it's FQDN where management should happen is better, or do you 'just' want to have the IP
> 
>     * configurable default group all hosts applied to
The current plugin I made has the following 'all' groups (where proxmox_ is a configurable prefix):

* proxmox_all_running
* proxmox_all_stopped
* proxmox_all_lxc
* proxmox_all_qemu
* proxmox_nodes

Given the above, do you still feel you need a group that has all of these in it?

> The main advantage of using libraries is the abstraction and potential duplicate code. Lets say there are 5 proxmox modules and 2 other Plugins and everyone is writing own API requests, every of them is duplicating the same code + everyone need to care of API changes

True, but the Proxmox API itself is already an abstraction for the underlying LXC/QEMU services, and is (in my experience) quite stable, they have to be, as their own WUI also uses it :-)

But I'm not experienced enough as a programmer to fully weigh complexity (using the library) vs duplication (not using it). Mostly, because I don't know how and if the Proxmoxer API would change a lot if the Proxmox API itself would drastically change.

**Comment by xoxys** [comment]
There might be situations where you need to keep your inventory small e.g. using Ansible Tower with a limided Hosts license. Personally I want a clean inventory and I have machines (as template source) that I never want in my inventory. So I would need an option to exclude. Anyway, everything I need is already in my PR. I'll move this to a custom collection for now.

**Comment by xoxys** [comment]
PR could be closed

File: plugins/inventory/proxmox.py [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+    description:
+        - Get inventory hosts from the proxmox service.
+        - "Uses a configuration file as an inventory source, it must end in C(.proxmox.yml) or C(.proxmox.yaml) and has a C(plugin: proxmox) entry."
+    extends_documentation_fragment:
+        - inventory_cache
+    options:
+      plugin:
+        description: The name of this plugin, it should always be set to C(proxmox) for this plugin to recognize it as it's own.
+        required: yes
+        choices: ['proxmox']
```
Comment (felixfontein) [inline plugins/inventory/proxmox.py:25]:
```suggestion
        choices: ['community.general.proxmox']
```

**Comment by felixfontein** [comment]
> PR could be closed

Feel free to close it yourself.

**Pull request closed without merging**